### PR TITLE
[KYUUBI #1059] Add Plan Only Operations

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

Introduce a session-level configuration `operation.plan.only.mode` to change ExecuteStatement to PlanOnlyStatement,

PlanOnlyStatement will only parse, analyze or optimize the statements instead of executing them.

This feature could run all Spark SQLs to verify the semantic compatibility without having side effects, which is very useful for testing, migrating without query rewrite.

This is similar to the `explain` command

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @pan3793 @ulysses-you @yanghua @hddong @turboFei

**Review by ulysses-you — APPROVED** [review]

**Comment by cfmcgrady** [comment]
Should we also log statement events when the `operation.plan.only.mode` is not `None`?

**Comment by ulysses-you** [comment]
> Should we also log statement events when the operation.plan.only.mode is not None?

Not see the requirements, this operation is more similar with`GetSchemas` etc ..

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala [hunk 1]
```diff
@@ -58,7 +61,15 @@ class SparkSQLOperationManager private (name: String) extends OperationManager(n
       runAsync: Boolean,
       queryTimeout: Long): Operation = {
     val spark = getSparkSession(session.handle)
-    val operation = new ExecuteStatement(spark, session, statement, runAsync, queryTimeout)
+
+    val operationMode = {
+      val modeStr = spark.conf.get(OPERATION_PLAN_ONLY.key, NONE.toString).toUpperCase(Locale.ROOT)
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:66]: `.toUpperCase(Locale.ROOT)` seems not necessary

File: kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala [hunk 2]
```diff
@@ -379,4 +379,31 @@ trait JDBCTests extends BasicJDBCTests {
...
+
+  test("KYUUBI #1059: Plan only operations") {
+    val ddl = "create table t(a int) using parquet"
+    val dql = "select * from t"
+    withJdbcStatement("t") { statement =>
+      try {
+        statement.execute("SET kyuubi.operation.plan.only.mode=optimize")
+        val set = statement.executeQuery(ddl)
+        assert(set.next())
+        assert(set.getString("plan") startsWith "Create")
```
Comment (pan3793) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala:391]: Is the result stable? If not stable, can we paste a demo result here?

File: kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala [hunk 3]
```diff
@@ -379,4 +379,31 @@ trait JDBCTests extends BasicJDBCTests {
...
+        val e1 = intercept[SQLException](statement.executeQuery(dql))
+        assert(e1.getMessage.contains("Table or view not found"))
+        statement.execute("SET kyuubi.operation.plan.only.mode=analyze")
+        val e2 = intercept[SQLException](statement.executeQuery(dql))
+        assert(e2.getMessage.contains("Table or view not found"))
+        statement.execute("SET kyuubi.operation.plan.only.mode=parse")
+        val set1 = statement.executeQuery(dql)
+        assert(set1.next())
+        assert(set1.getString("plan") contains "Unresolved")
```
Comment (pan3793) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala:400]: ditto

**Comment by yanghua** [comment]
LGTM

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:66]: we are getting it from spark runtime config

File: kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/operation/JDBCTests.scala:391]: If assert fails, we can see the original plan

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala (hunk 1 above)
Comment (hddong) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:66]: If we set a default value other than `None` in kyuubi server, seems the setting will not work here.

**Review by pan3793 — APPROVED** [review]
LGTM

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala]: updated

**Comment by hddong** [comment]
LGTM

File: kyuubi-server/src/test/scala/org/apache/kyuubi/operation/PlanOnlyOperationSuite.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (timothy65535) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/operation/PlanOnlyOperationSuite.scala:1]: `PlanOnlyOperationSuite.scala` to `KyuubiPlanOnlyOperationSuite.scala` will be better.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala [hunk 5]
```diff
@@ -52,13 +55,21 @@ class SparkSQLOperationManager private (name: String) extends OperationManager(n
...
       statement: String,
       runAsync: Boolean,
       queryTimeout: Long): Operation = {
     val spark = getSparkSession(session.handle)
-    val operation = new ExecuteStatement(spark, session, statement, runAsync, queryTimeout)
+
+    val operationModeStr =
+      spark.conf.get(OPERATION_PLAN_ONLY.key, operationModeDefault).toUpperCase(Locale.ROOT)
+    val operation = OperationModes.withName(operationModeStr) match {
+      case NONE => new ExecuteStatement(spark, session, statement, runAsync, queryTimeout)
```
Comment (timothy65535) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:70]: nit: do we need to consider the following case?
```
ser kyuubi.operation.plan.only.mode=abc;
select 1;
```

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:70]: would you like to make a follow up？

Comment (timothy65535) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:70]: Sorry, don't want to do.

File: kyuubi-server/src/test/scala/org/apache/kyuubi/operation/PlanOnlyOperationSuite.scala (hunk 4 above)
Comment (timothy65535) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/operation/PlanOnlyOperationSuite.scala:1]: ?

**Pull request closed without merging**